### DISCUSSION TITLE
TypeORM with Postgres docker

### DISCUSSION THREAD

**Discussion by boukmi** [jmcdo29/testing-nestjs#1677] [Q&A]
Hi everyone,

First of all, thank you to the author of this repository, it is really helpful.
I took a look at the TypeORM sample and it looks like we need to manually start the docker image to be able to run the integration tests.

I guess my question is, is there a way to make NestJS automatically spin up the docker image during the test phase?

Thank you

**Comment by jmcdo29** [comment]
Would you like to do this purely with js/ts code, or would you like to use shell scripts as a part of the test flow?

**Reply by boukmi** [reply]
Hi Jay,
js/ts would the preferable solution for us, but I'm open to other solutions that will help accomplish this.

Thanks

**Reply by jmcdo29** [reply]
I've got some code [in this repo here](https://github.com/jmcdo29/nest-e2e-sample/blob/main/test/index.spec.ts) that fires up a `docker compose` instance if it's not already running. You could adapt this to be in a `globalSetup` file for `jest` if you wanted. 

```ts
const runPostgres = async () => {
  ogma.debug('Checking if Postgres is running');
  const dockerStatus = JSON.parse(
    (await exec('docker compose -p test ps --format json')).stdout.toString(),
  );
  if (dockerStatus[0]?.State === 'running') {
    return;
  }
  ogma.debug('Postgres is not running. Starting via docker compose');
  await exec('docker compose -p test -f docker-compose.test.yml up -d');
  let exitCode = 1;
  do {
    try {
      await exec('nc -z localhost 35432');
      const exit = (await exec('echo $?')).stdout.toString();
      exitCode = parseInt(exit);
    } catch {
      exitCode = 1;
    }
  } while (exitCode !== 0);
  ogma.debug('netcat returned code 0. Finalizing wait time');
  await new Promise((resolve) => setTimeout(resolve, 1000));
};
```

**Reply by boukmi** [reply]
Thank you Jay.
I'll give this a shot and hopefully this will work.

Thank you again, I really appreciate it.

**Comment by barraponto** [comment]
You should look into [TestContainers](https://testcontainers.com/guides/getting-started-with-testcontainers-for-nodejs/)

**Comment by TatyOko28** [comment]
Yes, you can automate spinning up a Docker container for Postgres during the test phase using tools like Testcontainers, Docker Compose, or by scripting it directly in your test setup.

✅ Best Option: Use Testcontainers to programmatically manage the lifecycle of Docker containers during your tests.
✅ Alternative: Use Docker Compose and manage the container lifecycle outside your application.
